core/commands/dns.go

```diff
@@ -63,7 +63,7 @@ The resolver will give:
 			depth = namesys.DefaultDepthLimit
 		}
 		output, err := resolver.ResolveN(req.Context().Context, name, depth)
-		if err != nil {
+		if err != nil && (err != namesys.ErrResolveRecursion || depth > 1) {
```

Comment:
On Wed, Jun 10, 2015 at 12:31:33PM -0700, Juan Batiz-Benet wrote:
> > -		if err != nil {
> > +		if err != nil && (err != namesys.ErrResolveRecursion || depth > 1) {
> 
> there's another way to look at it:
> 
> - `resolve X` means "dereference X"
> - `resolve --depth=3 X` means "dereference X 3 times"

I don't think this makes sense.  What if you can't dereference anymore
after 2 steps?  Is that an error?

> - `resolve --recursive X` means "dereference X until you cannot
    dereference any more"

This has an implicit limit on the maximum allowed depth [1,2].  Do you
expect it to have different error handling from ‘--depth 32’?  For
example, if the result after 32 steps is still in a mutable namespace,
should --recursive raise an error?  Should ‘--depth 32’ raise an
error?

> If the command is satisfied, that's not really an error. We may also
> have IPNS values that dont even point to `/ipfs/...`. I think it's
> up to the user to figure out what to do with the returned
> value. This is a simpler interface with less complexity (having to
> figure out when to check the error, or not, etc).

Are you suggesting we replace our current ResolveN interface [3] with:

  ResolveNAllowUnresolved(ctx context.Context, name string, depth int) (value path.Path, err error)
  ResolveNExpectResolved(ctx context.Context, name string, depth int) (value path.Path, err error)

(and likely similar for the implicit-depth-limited Resolve())?  With
the dissection being that hitting the depth limit returns no path and
ErrResolveRecursion for *ExpectResolved and the final path and no
error for *AllowUnresolved.  Maybe that seems more approachable, but
it means the “I'd like to try drilling deeper” case [4] has to start
over at the beginning (since it doesn't have access to the path
reached by *ExpectResolved).  Or if you were using AllowUnresolved,
you'd need an additional check in the caller to decide if the path you
got back was resolved or not.

[1]: https://github.com/ipfs/go-ipfs/blob/feba3e1da5d43684c949828c35e72256272fc7d8/core/commands/resolve.go#L78
[2]: https://github.com/ipfs/go-ipfs/blob/feba3e1da5d43684c949828c35e72256272fc7d8/namesys/interface.go#L41-L42
[3]: https://github.com/ipfs/go-ipfs/blob/feba3e1da5d43684c949828c35e72256272fc7d8/namesys/interface.go#L92-L99
[4]: https://github.com/ipfs/go-ipfs/pull/1351/files#r32135641